core/src/main/kotlin/xtdb/vector/IndirectMultiVectorReader.kt

```diff
@@ -0,0 +1,303 @@
+    }
+
+    companion object {
+        private val MERGE_FIELDS: IFn = requiringResolve("xtdb.types/merge-fields")
+        private val VEC_TO_WRITER: IFn = requiringResolve("xtdb.vector.writer/->writer")
+    }
+
+    init {
+        assert(readers.any { it != null })
+    }
+
+    private fun unsupported(): RuntimeException {
+        throw UnsupportedOperationException("IndirectMultiVectoReader")
+    }
+
+    private fun safeReader(idx: Int): IVectorReader {
+        return readers[readerIndirection[idx]] ?: throw unsupported()
+    }
+
+    override fun valueCount(): Int {
+        return readerIndirection.valueCount()
+    }
+
+    override fun getName(): String {
+        return readers.filterNotNull().first().name
+    }
+
+    override fun withName(colName: String): IVectorReader {
+        return RenamedVectorReader(this, colName)
+    }
+
+    override fun getField(): Field {
+        return vectorField
+    }
+
+    override fun hashCode(idx: Int, hasher: ArrowBufHasher): Int {
+        return safeReader(idx).hashCode(vectorIndirections[idx], hasher)
+    }
+
+    override fun isNull(idx: Int): Boolean {
+        val readerIdx = readerIndirection[idx]
+        return readerIdx < 0 || readers[readerIdx] == null || readers[readerIdx]!!.isNull(vectorIndirections[idx])
+    }
+
+    override fun getBoolean(idx: Int): Boolean {
+        return safeReader(idx).getBoolean(vectorIndirections[idx])
+    }
+
+    override fun getByte(idx: Int): Byte {
+        return safeReader(idx).getByte(vectorIndirections[idx])
+    }
+
+    override fun getShort(idx: Int): Short {
+        return safeReader(idx).getShort(vectorIndirections[idx])
+    }
+
+    override fun getInt(idx: Int): Int {
+        return safeReader(idx).getInt(vectorIndirections[idx])
+    }
+
+    override fun getLong(idx: Int): Long {
+        return safeReader(idx).getLong(vectorIndirections[idx])
+    }
+
+    override fun getFloat(idx: Int): Float {
+        return safeReader(idx).getFloat(vectorIndirections[idx])
+    }
+
+    override fun getDouble(idx: Int): Double {
+        return safeReader(idx).getDouble(vectorIndirections[idx])
+    }
+
+    override fun getBytes(idx: Int): ByteBuffer {
+        return safeReader(idx).getBytes(vectorIndirections[idx])
+    }
+
+    override fun getPointer(idx: Int): ArrowBufPointer {
+        return safeReader(idx).getPointer(vectorIndirections[idx])
+    }
+
+    override fun getPointer(idx: Int, reuse: ArrowBufPointer): ArrowBufPointer {
+        return safeReader(idx).getPointer(vectorIndirections[idx], reuse)
+    }
+
+    override fun getObject(idx: Int): Any? {
+        return safeReader(idx).getObject(vectorIndirections[idx])
+    }
+
+    override fun getObject(idx: Int, keyFn: IKeyFn<*>?): Any? {
+        return safeReader(idx).getObject(vectorIndirections[idx], keyFn)
+    }
+
+    override fun structKeyReader(colName: String): IVectorReader {
+        // TODO what happens if underlying readers have a different type?
```

Comment:
this is one we'll need to handle

Reply:
I think I put the `TODO` in when I started off. This should be handled fine by the recursive usage of IMVR.